domain/application/modelmigration/import.go

```diff
@@ -223,31 +225,546 @@ func (i *importOperation) makeAddress(addr description.Address) (*network.SpaceA
+		Provides:       provides,
+		Requires:       requires,
+		Peers:          peers,
+		ExtraBindings:  importExtraBindings(data.ExtraBindings()),
+		Storage:        storage,
+		Devices:        devices,
+		PayloadClasses: payloadClasses,
+		Containers:     containers,
+		Resources:      resources,
+	}, nil
+}
+
+func (i *importOperation) importCharmManifest(data description.CharmManifest) (*internalcharm.Manifest, error) {
+	// Officially the manifest is optional, so if we don't have any data, we can
+	// just return nil.
+	if data == nil {
+		return nil, nil
+	}
+
+	bases, err := importManifestBases(data.Bases())
+	if err != nil {
+		return nil, fmt.Errorf("import manifest bases: %w", err)
+	}
+	return &internalcharm.Manifest{
+		Bases: bases,
+	}, nil
+}
+
+func (i *importOperation) importCharmLXDProfile(data description.CharmMetadata) (*internalcharm.LXDProfile, error) {
+	// LXDProfile is optional, so if we don't have any data, we can just return
+	// nil. If it does exist, then it's JSON encoded blob that we need to
+	// unmarshal into the internalcharm.LXDProfile struct.
+
+	lxdProfile := data.LXDProfile()
+	if lxdProfile == "" {
+		return nil, nil
+	}
+
+	var profile internalcharm.LXDProfile
+	if err := json.Unmarshal([]byte(lxdProfile), &profile); err != nil {
+		return nil, fmt.Errorf("unmarshal lxd profile: %w", err)
+	}
+
+	return &profile, nil
+}
+
+func (i *importOperation) importCharmConfig(data description.CharmConfigs) (*internalcharm.Config, error) {
+	// Charm config is optional, so if we don't have any data, we can just
+	// return nil.
+	if data == nil {
+		return nil, nil
+	}
+
+	descriptionConfig := data.Configs()
+
+	options := make(map[string]internalcharm.Option, len(descriptionConfig))
+	for name, c := range descriptionConfig {
+		options[name] = internalcharm.Option{
+			Type:        c.Type(),
+			Description: c.Description(),
+			Default:     c.Default(),
+		}
+	}
+
+	return &internalcharm.Config{
+		Options: options,
+	}, nil
+
+}
+
+func (i *importOperation) importCharmActions(data description.CharmActions) (*internalcharm.Actions, error) {
+	// Charm actions is optional, so if we don't have any data, we can just
+	// return nil.
+	if data == nil {
+		return nil, nil
+	}
+
+	descriptionActions := data.Actions()
+
+	actions := make(map[string]internalcharm.ActionSpec, len(descriptionActions))
+	for name, a := range descriptionActions {
+		parameters, err := importCharmParameters(a.Parameters())
+		if err != nil {
+			return nil, fmt.Errorf("import charm parameters: %w", err)
+		}
+
+		actions[name] = internalcharm.ActionSpec{
+			Description:    a.Description(),
+			Parallel:       a.Parallel(),
+			ExecutionGroup: a.ExecutionGroup(),
+			Params:         parameters,
+		}
 	}
+
+	return &internalcharm.Actions{
+		ActionSpecs: actions,
+	}, nil
 }
 
-func (s stubCharm) Actions() *internalcharm.Actions {
-	return &internalcharm.Actions{}
+func importCharmUser(data description.CharmMetadata) (internalcharm.RunAs, error) {
```

Comment:
Some of the comments from the export file apply here as well, if you make changes there they should also be made here.